cpp/src/utilities/bit_util.cuh

```diff
@@ -37,80 +42,79 @@
 namespace gdf {
 namespace util {
 
-static constexpr int ValidSize = 32;
-using ValidType = uint32_t;
+template <typename T>
+constexpr inline std::size_t size_in_bits() { return sizeof(T) * CHAR_BIT; }
 
+template <typename T>
+constexpr inline std::size_t size_in_bits(const T&) { return size_in_bits<T>(); }
 
-// Instead of this function, use gdf_valid_allocation_size from legacy_bitmask.hpp
-//CUDA_HOST_DEVICE_CALLABLE
-//  size_t
-//  valid_size(size_t column_length)
-//{
-//  const size_t n_ints = (column_length / ValidSize) + ((column_length % ValidSize) ? 1 : 0);
-//  return n_ints * sizeof(ValidType);
-//}
+namespace detail {
 
-// Instead of this function, use gdf_is_valid from gdf/utils.h
-///CUDA_HOST_DEVICE_CALLABLE bool get_bit(const gdf_valid_type* const bits, size_t i)
-///{
-///  return  bits == nullptr? true :  bits[i >> size_t(3)] & (1 << (i & size_t(7)));
-///}
+template <typename BitContainer, typename Size>
+constexpr CUDA_HOST_DEVICE_CALLABLE
+Size intra_container_index(Size bit_index) { return bit_index % size_in_bits<BitContainer>(); }
 
-CUDA_HOST_DEVICE_CALLABLE
-  uint8_t
-  byte_bitmask(size_t i)
-{
-  static uint8_t kBitmask[8] = { 1, 2, 4, 8, 16, 32, 64, 128 };
-  return kBitmask[i];
-}
+template <typename BitContainer, typename Size>
+constexpr CUDA_HOST_DEVICE_CALLABLE
+Size bit_container_index(Size bit_index) { return bit_index / size_in_bits<BitContainer>(); }
+
+
+} // namespace detail
 
-CUDA_HOST_DEVICE_CALLABLE
-  uint8_t
-  flipped_bitmask(size_t i)
-{
-  static uint8_t kFlippedBitmask[] = { 254, 253, 251, 247, 239, 223, 191, 127 };
-  return kFlippedBitmask[i];
-}
 
-CUDA_HOST_DEVICE_CALLABLE void turn_bit_on(uint8_t* const bits, size_t i)
+template <typename BitContainer, typename Size>
+constexpr CUDA_HOST_DEVICE_CALLABLE
+void turn_bit_on(BitContainer* bits, Size bit_index)
```

Comment:
Again, the input is the record index, not the bit index.

Reply:
The functions in this file don't know about "columns" or "records". Specifically, this function gets a bunch of bits in memory, and an index into this sequence of bits.

Reply:
My mistake.